src/Magento/FunctionalTestingFramework/StaticCheck/DeprecatedEntityUsageCheck.php

```diff
@@ -0,0 +1,605 @@
+    {
+        $outArray = [];
+        foreach ($inArray as $key => $arr) {
+            $outArray[$key] = array_unique($arr);
+        }
+        return $outArray;
+    }
+
+    /**
+     * Return attribute value for each node in DOMNodeList as an array
+     *
+     * @param DOMNodeList $nodes
+     * @param string      $attributeName
+     * @return array
+     */
+    private function getAttributesFromDOMNodeList($nodes, $attributeName)
+    {
+        $attributes = [];
+        /** @var DOMElement $node */
+        foreach ($nodes as $node) {
+            $attributeValue = $node->getAttribute($attributeName);
+            if (!empty($attributeValue)) {
+                $attributes[] = $attributeValue;
+            }
+        }
+        return $attributes;
+    }
+
+    /**
+     * Find metadata from data array
+     *
+     * @param array  $references
+     * @param string $type
+     * @return array
+     */
+    private function getMetadataFromData($references, $type)
+    {
+        $metaDataReferences = [];
+        try {
+            foreach ($references as $dataName) {
+                /** @var EntityDataObject $dataEntity */
+                $dataEntity = $this->scriptUtil->findEntity($dataName);
+                if ($dataEntity) {
+                    $metadata = $dataEntity->getType();
+                    if (!empty($metadata)) {
+                        $metaDataReferences[$dataName][$metadata] = $type;
+                    }
+                }
+            }
+        } catch (Exception $e) {
+        }
+        return $metaDataReferences;
+    }
+
+    /**
+     * Find violating metadata references. Input array format is either
+     *
+     *  [
+     *      $dataName1 => [
+     *          $metaDataName1 = [
+     *              'create',
+     *              'update',
+     *          ],
+     *      ],
+     *      $dataName2 => [
+     *          $metaDataName2 => 'create',
+     *      ],
+     *      $dataName3 => [
+     *          $metaDataName3 = [
+     *              'get',
+     *              'create',
+     *              'update',
+     *          ],
+     *      ],
+     *      ...
+     *  ]
+     *
+     *  or
+     *
+     *  [
+     *      $dataName1 => [
+     *          $metaDataName1 => 'all',
+     *      ],
+     *      $dataName2 => [
+     *          $metaDataName2 => 'all',
+     *      ],
+     *      $dataName5 => [
+     *          $metaDataName5 => 'all',
+     *      ],
+     *      ...
+     *  ]
+     *
+     * @param array $references
+     * @return array
+     * @SuppressWarnings(PHPMD.CyclomaticComplexity)
+     * @SuppressWarnings(PHPMD.NPathComplexity)
+     */
+    private function findViolatingMetadataReferences($references)
+    {
+        $allObjects = OperationDefinitionObjectHandler::getInstance()->getAllObjects();
+
+        // Find Violations
+        $violatingReferences = [];
+        foreach ($references as $dataName => $metadataArray) {
```

Comment:
can $references capture the operation name eg: CreateCategory or UpdateCategory from CategoryMeta.xml. Currently the check only captures the dataType of the deprecated operation.
So in tests where both createData and updateData, the output only states one error -
eg test -
```
        <createData entity="SimpleSubCategory" stepKey="createCategory"/>
        <updateData entity="SimpleSubCategory" createDataKey="CategoryTitle" stepKey="updateCategory"/>
```
Output -
```
	- "SimpleSubCategory" references deprecated:
		"category" in metadata xml file
```

Reply:
Also, the error output seems a little unclear, may be update the output to this -
- Deprecated Metadata(s):
"CreateCategory" operation in metadata xml file for "SimpleSubCategory" data entity
"UpdateCategory" operation in metadata xml file for "SimpleSubCategory" data entity

Reply:
We are not always know which type of metadata in a violation. We only know when it's found in `<createData>`, etc. But we will know where it's referenced from, which is print out.